### ISSUE TITLE
switching locales no longer invalidates gettext translation cache since 5.28

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#17081]
Migrated from [rt.perl.org#134264](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134264) (status was 'open')

Searchable as RT134264$

**Comment by p5pRT** [comment]
### From @ntyni

#### Created by @ntyni



As discussed in https://bugs.debian.org/931139 and
https://bugs.debian.org/924657 glibc has a cache of already loaded gettext
translations that gets invalidated in setlocale\(3\)\, bindtextdomain\(3\)
and textdomain\(3\) but not uselocale\(3\)\.

Starting with Perl 5\.28\, Perl uses POSIX 2008 thread\-safe locales\, so
it calls uselocale\(3\) underneath when the Perl side POSIX&#8203;::setlocale\(\)
function is invoked\.

This makes gettext\(3\) think that a translation for the new locale is
already loaded when it really corresponds to the old locale\.

While this has resulted in a 5\.28 regression / behaviour change for Perl\,
it's not clear to me whether glibc is working correctly here or not\. Is
this something that Perl should or could work around somehow?

I'm attaching a test case in Perl using Locale&#8203;::gettext that outputs three
different strings on 5\.24\.1 \(Debian stretch\) but only one different string
\("lauseoppivirhe"\) three times on 5\.28\.1 \(Debian buster / sid\)\.

I'm also attaching a test case in C that demonstrates that the difference
is about using setlocale\(3\) vs uselocale\(3\)\.

Obviously the relevant locales \(fi\_FI\, fr\_FR\, en\_US\) need to be installed
on the system for these test cases to work\.

Please let me know if I'm not explaining this well enough\.
I'm having a bit of trouble choosing the level of detail here\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.28.1:

Configured by Debian at Sun Mar 31 11:51:22 UTC 2019.

Summary of my perl5 (revision 5 version 28 subversion 1) configuration:
   
  Platform:
    osname=linux
    osvers=4.9.0
    archname=x86_64-linux-gnu-thread-multi
    uname='linux localhost 4.9.0 #1 smp debian 4.9.0 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dcc=x86_64-linux-gnu-gcc -Dcpp=x86_64-linux-gnu-cpp -Dld=x86_64-linux-gnu-gcc -Dccflags=-DDEBIAN -Wdate-time -D_FORTIFY_SOURCE=2 -g -O2 -fdebug-prefix-map=/build/perl-5WfRyb/perl-5.28.1=. -fstack-protector-strong -Wformat -Werror=format-security -Dldflags= -Wl,-z,relro -Dlddlflags=-shared -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.28 -Darchlib=/usr/lib/x86_64-linux-gnu/perl/5.28 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/x86_64-linux-gnu/perl5/5.28 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.28.1 -Dsitearch=/usr/local/lib/x86_64-linux-gnu/perl/5.28.1 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl
-Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -Ui_xlocale -Uversiononly -DDEBUGGING=-g -Doptimize=-O2 -dEs -Duseshrplib -Dlibperl=libperl.so.5.28.1'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='x86_64-linux-gnu-gcc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='x86_64-linux-gnu-gcc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=libc-2.28.so
    so=so
    useshrplib=true
    libperl=libperl.so.5.28
    gnulibc_version='2.28'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -L/usr/local/lib -fstack-protector-strong'

Locally applied patches:
    DEBPKG:debian/cpan_definstalldirs - Provide a sensible INSTALLDIRS default for modules installed from CPAN.
    DEBPKG:debian/db_file_ver - https://bugs.debian.org/340047 Remove overly restrictive DB_File version check.
    DEBPKG:debian/doc_info - Replace generic man(1) instructions with Debian-specific information.
    DEBPKG:debian/enc2xs_inc - https://bugs.debian.org/290336 Tweak enc2xs to follow symlinks and ignore missing @INC directories.
    DEBPKG:debian/errno_ver - https://bugs.debian.org/343351 Remove Errno version check due to upgrade problems with long-running processes.
    DEBPKG:debian/libperl_embed_doc - https://bugs.debian.org/186778 Note that libperl-dev package is required for embedded linking
    DEBPKG:fixes/respect_umask - Respect umask during installation
    DEBPKG:debian/writable_site_dirs - Set umask approproately for site install directories
    DEBPKG:debian/extutils_set_libperl_path - EU:MM: set location of libperl.a under /usr/lib
    DEBPKG:debian/no_packlist_perllocal - Don't install .packlist or perllocal.pod for perl or vendor
    DEBPKG:debian/fakeroot - Postpone LD_LIBRARY_PATH evaluation to the binary targets.
    DEBPKG:debian/instmodsh_doc - Debian policy doesn't install .packlist files for core or vendor.
    DEBPKG:debian/ld_run_path - Remove standard libs from LD_RUN_PATH as per Debian policy.
    DEBPKG:debian/libnet_config_path - Set location of libnet.cfg to /etc/perl/Net as /usr may not be writable.
    DEBPKG:debian/perlivp - https://bugs.debian.org/510895 Make perlivp skip include directories in /usr/local
    DEBPKG:debian/squelch-locale-warnings - https://bugs.debian.org/508764 Squelch locale warnings in Debian package maintainer scripts
    DEBPKG:debian/patchlevel - https://bugs.debian.org/567489 List packaged patches for 5.28.1-6 in patchlevel.h
    DEBPKG:fixes/document_makemaker_ccflags - https://bugs.debian.org/628522 [rt.cpan.org #68613] Document that CCFLAGS should include $Config{ccflags}
    DEBPKG:debian/find_html2text - https://bugs.debian.org/640479 Configure CPAN::Distribution with correct name of html2text
    DEBPKG:debian/perl5db-x-terminal-emulator.patch - https://bugs.debian.org/668490 Invoke x-terminal-emulator rather than xterm in perl5db.pl
    DEBPKG:debian/cpan-missing-site-dirs - https://bugs.debian.org/688842 Fix CPAN::FirstTime defaults with nonexisting site dirs if a parent is writable
    DEBPKG:fixes/memoize_storable_nstore - [rt.cpan.org #77790] https://bugs.debian.org/587650 Memoize::Storable: respect 'nstore' option not respected
    DEBPKG:debian/makemaker-pasthru - https://bugs.debian.org/758471 Pass LD settings through to subdirectories
    DEBPKG:debian/makemaker-manext - https://bugs.debian.org/247370 Make EU::MakeMaker honour MANnEXT settings in generated manpage headers
    DEBPKG:debian/kfreebsd-softupdates - https://bugs.debian.org/796798 Work around Debian Bug#796798
    DEBPKG:fixes/autodie-scope - https://bugs.debian.org/798096 Fix a scoping issue with "no autodie" and the "system" sub
    DEBPKG:fixes/memoize-pod - [rt.cpan.org #89441] Fix POD errors in Memoize
    DEBPKG:debian/hurd-softupdates - https://bugs.debian.org/822735 Fix t/op/stat.t failures on hurd
    DEBPKG:fixes/math_complex_doc_great_circle - https://bugs.debian.org/697567 [rt.cpan.org #114104] Math::Trig: clarify definition of great_circle_midpoint
    DEBPKG:fixes/math_complex_doc_see_also - https://bugs.debian.org/697568 [rt.cpan.org #114105] Math::Trig: add missing SEE ALSO
    DEBPKG:fixes/math_complex_doc_angle_units - https://bugs.debian.org/731505 [rt.cpan.org #114106] Math::Trig: document angle units
    DEBPKG:fixes/cpan_web_link - https://bugs.debian.org/367291 CPAN: Add link to main CPAN web site
    DEBPKG:debian/hppa_op_optimize_workaround - https://bugs.debian.org/838613 Temporarily lower the optimization of op.c on hppa due to gcc-6 problems
    DEBPKG:debian/installman-utf8 - https://bugs.debian.org/840211 Generate man pages with UTF-8 characters
    DEBPKG:fixes/getopt-long-4 - https://bugs.debian.org/864544 [rt.cpan.org #122068] Fix issue #122068.
    DEBPKG:debian/hppa_opmini_optimize_workaround - https://bugs.debian.org/869122 Lower the optimization level of opmini.c on hppa
    DEBPKG:debian/sh4_op_optimize_workaround - https://bugs.debian.org/869373 Also lower the optimization level of op.c and opmini.c on sh4
    DEBPKG:debian/perldoc-pager - https://bugs.debian.org/870340 [rt.cpan.org #120229] Fix perldoc terminal escapes when sensible-pager is less
    DEBPKG:debian/prune_libs - https://bugs.debian.org/128355 Prune the list of libraries wanted to what we actually need.
    DEBPKG:debian/mod_paths - Tweak @INC ordering for Debian
    DEBPKG:debian/configure-regen - https://bugs.debian.org/762638 Regenerate Configure et al. after probe unit changes
    DEBPKG:debian/deprecate-with-apt - https://bugs.debian.org/747628 Point users to Debian packages of deprecated core modules
    DEBPKG:debian/disable-stack-check - https://bugs.debian.org/902779 [perl #133327] Disable debugperl stack extension checks for binary compatibility with perl
    DEBPKG:debian/gdbm-fatal - [perl #133295] https://bugs.debian.org/904005 Temporarily skip GDBM_File fatal.t for gdbm >= 1.15 compatibility
    DEBPKG:fixes/storable-recursion - https://bugs.debian.org/912900 [perl #133326] [120060c] (perl #133326) fix and clarify handling of recurs_sv.
    DEBPKG:fixes/caretx-fallback - https://bugs.debian.org/913347 [perl #133573] [03b94aa] RT#133573: $^X fallback when platform-specific technique fails
    DEBPKG:fixes/eumm-usrmerge - https://bugs.debian.org/913637 Avoid mangling /bin non-perl shebangs on merged-/usr systems
    DEBPKG:fixes/errno-include-path - [6c5080f] [perl #133662] https://bugs.debian.org/875921 Make Errno_pm.PL compatible with /usr/include/<ARCH>/errno.h
    DEBPKG:fixes/kfreebsd-renameat - [a3c63a9] https://bugs.debian.org/912521 [perl #133668] Also work around renameat() kernel bug on GNU/kFreeBSD
    DEBPKG:fixes/time-local-2020 - https://bugs.debian.org/915209 [rt.cpan.org #124787] Fix Time::Local tests
    DEBPKG:fixes/inplace-editing-bugfix/part1 - https://bugs.debian.org/914651 (perl #133659) move argvout cleanup to a new function
    DEBPKG:fixes/inplace-editing-bugfix/part2 - https://bugs.debian.org/914651 (perl #133659) tests for global destruction handling of inplace editing
    DEBPKG:fixes/inplace-editing-bugfix/part3 - https://bugs.debian.org/914651 (perl #133659) make an in-place edit successful if the exit status is zero
    DEBPKG:fixes/fix-manifest-failures - https://bugs.debian.org/914962 Fix t/porting/manifest.t failures when run in a foreign git checkout
    DEBPKG:fixes/pipe-open-bugfix/part1 - [perl #133726] https://bugs.debian.org/916313 Always mark pipe in pipe-open as inherit-on-exec
    DEBPKG:fixes/pipe-open-bugfix/part2 - [perl #133726] https://bugs.debian.org/916313 Always mark pipe in list pipe-open as inherit-on-exec
    DEBPKG:fixes/storable-probing/prereq1 - [3f4cad1] Storable: fix for strawberry build failures:
    DEBPKG:fixes/storable-probing/prereq2 - [perl #133411] [edf639f] (perl #133411) don't try to load Storable with -Dusecrosscompile
    DEBPKG:fixes/storable-probing/disable-probing - https://bugs.debian.org/914133 [perl #133708] [2a0bbd3] (perl #133708) remove build-time probing for stack limits for Storable
    DEBPKG:debian/perlbug-editor - https://bugs.debian.org/922609 Use "editor" as the default perlbug editor, as per Debian policy
    DEBPKG:fixes/posix-mbrlen - [25d7b7a] https://bugs.debian.org/924517 [perl #133928] Fix POSIX::mblen mbstate_t initialization on threaded perls with glibc


@INC for perl 5.28.1:
    /etc/perl
    /usr/local/lib/x86_64-linux-gnu/perl/5.28.1
    /usr/local/share/perl/5.28.1
    /usr/lib/x86_64-linux-gnu/perl5/5.28
    /usr/share/perl5
    /usr/lib/x86_64-linux-gnu/perl/5.28
    /usr/share/perl/5.28
    /usr/local/lib/site_perl
    /usr/lib/x86_64-linux-gnu/perl-base


Environment for perl 5.28.1:
    HOME=/home/niko
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_CTYPE=fi_FI.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/niko/bin:/home/niko/bin:/home/niko/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/sbin:/usr/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @ntyni

[931139.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1645179/839759/931139.pl)

**Comment by p5pRT** [comment]
### From @ntyni


```C
#include <stdio.h>
#include <stdlib.h>

#include <libintl.h>
#include <locale.h>

/* https://bugs.debian.org/931139 */

int main(int argc, char **argv)
{
  locale_t loc;
  int i=0;

  /* The C locale is special cased in glibc to not look at LANGUAGE
     so we set C.UTF-8 as the base locale instead */
  setenv("LANG", "C.UTF-8", 1);

  char *locales[] = { "fi_FI", "fr_FR", "en_US", NULL };

  for (i=0; locales[i] != NULL; i++) {
    setenv("LANGUAGE", locales[i], 1);
    if (argc > 1) {
        /* "Old" behaviour, no bug */
        setlocale(LC_MESSAGES, "");
    } else {
        /* "New" behaviour, first translation stays cached */
        loc = newlocale(LC_MESSAGES_MASK, "", (locale_t) 0);
        if (loc == (locale_t) 0)
            perror("newlocale");
        uselocale(loc);
    }
    printf("%s\n", dgettext("bash", "syntax error"));
  }

  return EXIT_SUCCESS;
}

```

**Comment by p5pRT** [comment]
### From @khwilliamson

On 7/4/19 12&#8203;:01 PM\, Niko Tyni \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Niko Tyni
> \# Please include the string&#8203;:  \[perl \#134264\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=134264 >
> 
> 
> This is a bug report for perl from ntyni@&#8203;debian\.org\,
> generated with the help of perlbug 1\.41 running under perl 5\.28\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> As discussed in https://bugs.debian.org/931139 and
> https://bugs.debian.org/924657 glibc has a cache of already loaded gettext
> translations that gets invalidated in setlocale\(3\)\, bindtextdomain\(3\)
> and textdomain\(3\) but not uselocale\(3\)\.
> 
> Starting with Perl 5\.28\, Perl uses POSIX 2008 thread\-safe locales\, so
> it calls uselocale\(3\) underneath when the Perl side POSIX&#8203;::setlocale\(\)
> function is invoked\.
> 
> This makes gettext\(3\) think that a translation for the new locale is
> already loaded when it really corresponds to the old locale\.
> 
> While this has resulted in a 5\.28 regression / behaviour change for Perl\,
> it's not clear to me whether glibc is working correctly here or not\. Is
> this something that Perl should or could work around somehow?

This looks like an oversight in glibc to me\, hence their bug\.  No matter 
how the locale gets changed\, the cache should be invalidated\.

In looking at the links you provided\, I didn't see how to read the 
proposed patch\.

We should look for workarounds in case glibc refuses to change \(I don't 
have respect for their willingness to acknowledge flaws in their code\)\, 
or for working on earlier versions than any fix they do make\.  I wonder 
if calling foo=textdomain\(NULL\);textdomain\(foo\);
would do it\.
> 
> I'm attaching a test case in Perl using Locale&#8203;::gettext that outputs three
> different strings on 5\.24\.1 \(Debian stretch\) but only one different string
> \("lauseoppivirhe"\) three times on 5\.28\.1 \(Debian buster / sid\)\.
> 
> I'm also attaching a test case in C that demonstrates that the difference
> is about using setlocale\(3\) vs uselocale\(3\)\.
> 
> Obviously the relevant locales \(fi\_FI\, fr\_FR\, en\_US\) need to be installed
> on the system for these test cases to work\.
> 
> Please let me know if I'm not explaining this well enough\.
> I'm having a bit of trouble choosing the level of detail here\.
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>      category=core
>      severity=medium
> \-\-\-
> Site configuration information for perl 5\.28\.1&#8203;:
> 
> Configured by Debian at Sun Mar 31 11&#8203;:51&#8203;:22 UTC 2019\.
> 
> Summary of my perl5 \(revision 5 version 28 subversion 1\) configuration&#8203;:
>     
>    Platform&#8203;:
>      osname=linux
>      osvers=4\.9\.0
>      archname=x86\_64\-linux\-gnu\-thread\-multi
>      uname='linux localhost 4\.9\.0 \#1 smp debian 4\.9\.0 x86\_64 gnulinux '
>      config\_args='\-Dusethreads \-Duselargefiles \-Dcc=x86\_64\-linux\-gnu\-gcc \-Dcpp=x86\_64\-linux\-gnu\-cpp \-Dld=x86\_64\-linux\-gnu\-gcc \-Dccflags=\-DDEBIAN \-Wdate\-time \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fdebug\-prefix\-map=/build/perl\-5WfRyb/perl\-5\.28\.1=\. \-fstack\-protector\-strong \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.28 \-Darchlib=/usr/lib/x86\_64\-linux\-gnu/perl/5\.28 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/x86\_64\-linux\-gnu/perl5/5\.28 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.28\.1 \-Dsitearch=/usr/local/lib/x86\_64\-linux\-gnu/perl/5\.28\.1 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl
> \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-Ui\_xlocale \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-dEs \-Duseshrplib \-Dlibperl=libperl\.so\.5\.28\.1'
>      hint=recommended
>      useposix=true
>      d\_sigaction=define
>      useithreads=define
>      usemultiplicity=define
>      use64bitint=define
>      use64bitall=define
>      uselongdouble=undef
>      usemymalloc=n
>      default\_inc\_excludes\_dot=define
>      bincompat5005=undef
>    Compiler&#8203;:
>      cc='x86\_64\-linux\-gnu\-gcc'
>      ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
>      optimize='\-O2 \-g'
>      cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
>      ccversion=''
>      gccversion='8\.3\.0'
>      gccosandvers=''
>      intsize=4
>      longsize=8
>      ptrsize=8
>      doublesize=8
>      byteorder=12345678
>      doublekind=3
>      d\_longlong=define
>      longlongsize=8
>      d\_longdbl=define
>      longdblsize=16
>      longdblkind=3
>      ivtype='long'
>      ivsize=8
>      nvtype='double'
>      nvsize=8
>      Off\_t='off\_t'
>      lseeksize=8
>      alignbytes=8
>      prototype=define
>    Linker and Libraries&#8203;:
>      ld='x86\_64\-linux\-gnu\-gcc'
>      ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
>      libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/8/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
>      libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
>      perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
>      libc=libc\-2\.28\.so
>      so=so
>      useshrplib=true
>      libperl=libperl\.so\.5\.28
>      gnulibc\_version='2\.28'
>    Dynamic Linking&#8203;:
>      dlsrc=dl\_dlopen\.xs
>      dlext=so
>      d\_dlsymun=undef
>      ccdlflags='\-Wl\,\-E'
>      cccdlflags='\-fPIC'
>      lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector\-strong'
> 
> Locally applied patches&#8203;:
>      DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
>      DEBPKG&#8203;:debian/db\_file\_ver \- https://bugs.debian.org/340047 Remove overly restrictive DB\_File version check\.
>      DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
>      DEBPKG&#8203;:debian/enc2xs\_inc \- https://bugs.debian.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
>      DEBPKG&#8203;:debian/errno\_ver \- https://bugs.debian.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
>      DEBPKG&#8203;:debian/libperl\_embed\_doc \- https://bugs.debian.org/186778 Note that libperl\-dev package is required for embedded linking
>      DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
>      DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site install directories
>      DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: set location of libperl\.a under /usr/lib
>      DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or perllocal\.pod for perl or vendor
>      DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
>      DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
>      DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
>      DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
>      DEBPKG&#8203;:debian/perlivp \- https://bugs.debian.org/510895 Make perlivp skip include directories in /usr/local
>      DEBPKG&#8203;:debian/squelch\-locale\-warnings \- https://bugs.debian.org/508764 Squelch locale warnings in Debian package maintainer scripts
>      DEBPKG&#8203;:debian/patchlevel \- https://bugs.debian.org/567489 List packaged patches for 5\.28\.1\-6 in patchlevel\.h
>      DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- https://bugs.debian.org/628522 \[rt\.cpan\.org \#68613\] Document that CCFLAGS should include $Config\{ccflags\}
>      DEBPKG&#8203;:debian/find\_html2text \- https://bugs.debian.org/640479 Configure CPAN&#8203;::Distribution with correct name of html2text
>      DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- https://bugs.debian.org/668490 Invoke x\-terminal\-emulator rather than xterm in perl5db\.pl
>      DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- https://bugs.debian.org/688842 Fix CPAN&#8203;::FirstTime defaults with nonexisting site dirs if a parent is writable
>      DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] https://bugs.debian.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' option not respected
>      DEBPKG&#8203;:debian/makemaker\-pasthru \- https://bugs.debian.org/758471 Pass LD settings through to subdirectories
>      DEBPKG&#8203;:debian/makemaker\-manext \- https://bugs.debian.org/247370 Make EU&#8203;::MakeMaker honour MANnEXT settings in generated manpage headers
>      DEBPKG&#8203;:debian/kfreebsd\-softupdates \- https://bugs.debian.org/796798 Work around Debian Bug\#796798
>      DEBPKG&#8203;:fixes/autodie\-scope \- https://bugs.debian.org/798096 Fix a scoping issue with "no autodie" and the "system" sub
>      DEBPKG&#8203;:fixes/memoize\-pod \- \[rt\.cpan\.org \#89441\] Fix POD errors in Memoize
>      DEBPKG&#8203;:debian/hurd\-softupdates \- https://bugs.debian.org/822735 Fix t/op/stat\.t failures on hurd
>      DEBPKG&#8203;:fixes/math\_complex\_doc\_great\_circle \- https://bugs.debian.org/697567 \[rt\.cpan\.org \#114104\] Math&#8203;::Trig&#8203;: clarify definition of great\_circle\_midpoint
>      DEBPKG&#8203;:fixes/math\_complex\_doc\_see\_also \- https://bugs.debian.org/697568 \[rt\.cpan\.org \#114105\] Math&#8203;::Trig&#8203;: add missing SEE ALSO
>      DEBPKG&#8203;:fixes/math\_complex\_doc\_angle\_units \- https://bugs.debian.org/731505 \[rt\.cpan\.org \#114106\] Math&#8203;::Trig&#8203;: document angle units
>      DEBPKG&#8203;:fixes/cpan\_web\_link \- https://bugs.debian.org/367291 CPAN&#8203;: Add link to main CPAN web site
>      DEBPKG&#8203;:debian/hppa\_op\_optimize\_workaround \- https://bugs.debian.org/838613 Temporarily lower the optimization of op\.c on hppa due to gcc\-6 problems
>      DEBPKG&#8203;:debian/installman\-utf8 \- https://bugs.debian.org/840211 Generate man pages with UTF\-8 characters
>      DEBPKG&#8203;:fixes/getopt\-long\-4 \- https://bugs.debian.org/864544 \[rt\.cpan\.org \#122068\] Fix issue \#122068\.
>      DEBPKG&#8203;:debian/hppa\_opmini\_optimize\_workaround \- https://bugs.debian.org/869122 Lower the optimization level of opmini\.c on hppa
>      DEBPKG&#8203;:debian/sh4\_op\_optimize\_workaround \- https://bugs.debian.org/869373 Also lower the optimization level of op\.c and opmini\.c on sh4
>      DEBPKG&#8203;:debian/perldoc\-pager \- https://bugs.debian.org/870340 \[rt\.cpan\.org \#120229\] Fix perldoc terminal escapes when sensible\-pager is less
>      DEBPKG&#8203;:debian/prune\_libs \- https://bugs.debian.org/128355 Prune the list of libraries wanted to what we actually need\.
>      DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
>      DEBPKG&#8203;:debian/configure\-regen \- https://bugs.debian.org/762638 Regenerate Configure et al\. after probe unit changes
>      DEBPKG&#8203;:debian/deprecate\-with\-apt \- https://bugs.debian.org/747628 Point users to Debian packages of deprecated core modules
>      DEBPKG&#8203;:debian/disable\-stack\-check \- https://bugs.debian.org/902779 \[perl \#133327\] Disable debugperl stack extension checks for binary compatibility with perl
>      DEBPKG&#8203;:debian/gdbm\-fatal \- \[perl \#133295\] https://bugs.debian.org/904005 Temporarily skip GDBM\_File fatal\.t for gdbm >= 1\.15 compatibility
>      DEBPKG&#8203;:fixes/storable\-recursion \- https://bugs.debian.org/912900 \[perl \#133326\] \[120060c\] \(perl \#133326\) fix and clarify handling of recurs\_sv\.
>      DEBPKG&#8203;:fixes/caretx\-fallback \- https://bugs.debian.org/913347 \[perl \#133573\] \[03b94aa\] RT\#133573&#8203;: $^X fallback when platform\-specific technique fails
>      DEBPKG&#8203;:fixes/eumm\-usrmerge \- https://bugs.debian.org/913637 Avoid mangling /bin non\-perl shebangs on merged\-/usr systems
>      DEBPKG&#8203;:fixes/errno\-include\-path \- \[6c5080f\] \[perl \#133662\] https://bugs.debian.org/875921 Make Errno\_pm\.PL compatible with /usr/include/\<ARCH>/errno\.h
>      DEBPKG&#8203;:fixes/kfreebsd\-renameat \- \[a3c63a9\] https://bugs.debian.org/912521 \[perl \#133668\] Also work around renameat\(\) kernel bug on GNU/kFreeBSD
>      DEBPKG&#8203;:fixes/time\-local\-2020 \- https://bugs.debian.org/915209 \[rt\.cpan\.org \#124787\] Fix Time&#8203;::Local tests
>      DEBPKG&#8203;:fixes/inplace\-editing\-bugfix/part1 \- https://bugs.debian.org/914651 \(perl \#133659\) move argvout cleanup to a new function
>      DEBPKG&#8203;:fixes/inplace\-editing\-bugfix/part2 \- https://bugs.debian.org/914651 \(perl \#133659\) tests for global destruction handling of inplace editing
>      DEBPKG&#8203;:fixes/inplace\-editing\-bugfix/part3 \- https://bugs.debian.org/914651 \(perl \#133659\) make an in\-place edit successful if the exit status is zero
>      DEBPKG&#8203;:fixes/fix\-manifest\-failures \- https://bugs.debian.org/914962 Fix t/porting/manifest\.t failures when run in a foreign git checkout
>      DEBPKG&#8203;:fixes/pipe\-open\-bugfix/part1 \- \[perl \#133726\] https://bugs.debian.org/916313 Always mark pipe in pipe\-open as inherit\-on\-exec
>      DEBPKG&#8203;:fixes/pipe\-open\-bugfix/part2 \- \[perl \#133726\] https://bugs.debian.org/916313 Always mark pipe in list pipe\-open as inherit\-on\-exec
>      DEBPKG&#8203;:fixes/storable\-probing/prereq1 \- \[3f4cad1\] Storable&#8203;: fix for strawberry build failures&#8203;:
>      DEBPKG&#8203;:fixes/storable\-probing/prereq2 \- \[perl \#133411\] \[edf639f\] \(perl \#133411\) don't try to load Storable with \-Dusecrosscompile
>      DEBPKG&#8203;:fixes/storable\-probing/disable\-probing \- https://bugs.debian.org/914133 \[perl \#133708\] \[2a0bbd3\] \(perl \#133708\) remove build\-time probing for stack limits for Storable
>      DEBPKG&#8203;:debian/perlbug\-editor \- https://bugs.debian.org/922609 Use "editor" as the default perlbug editor\, as per Debian policy
>      DEBPKG&#8203;:fixes/posix\-mbrlen \- \[25d7b7a\] https://bugs.debian.org/924517 \[perl \#133928\] Fix POSIX&#8203;::mblen mbstate\_t initialization on threaded perls with glibc
> 
> \-\-\-
> @&#8203;INC for perl 5\.28\.1&#8203;:
>      /etc/perl
>      /usr/local/lib/x86\_64\-linux\-gnu/perl/5\.28\.1
>      /usr/local/share/perl/5\.28\.1
>      /usr/lib/x86\_64\-linux\-gnu/perl5/5\.28
>      /usr/share/perl5
>      /usr/lib/x86\_64\-linux\-gnu/perl/5\.28
>      /usr/share/perl/5\.28
>      /usr/local/lib/site\_perl
>      /usr/lib/x86\_64\-linux\-gnu/perl\-base
> 
> \-\-\-
> Environment for perl 5\.28\.1&#8203;:
>      HOME=/home/niko
>      LANG=en\_US\.UTF\-8
>      LANGUAGE \(unset\)
>      LC\_CTYPE=fi\_FI\.UTF\-8
>      LD\_LIBRARY\_PATH \(unset\)
>      LOGDIR \(unset\)
>      PATH=/home/niko/bin&#8203;:/home/niko/bin&#8203;:/home/niko/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/local/games&#8203;:/usr/games&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/sbin
>      PERL\_BADLANG \(unset\)
>      SHELL=/bin/zsh
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ntyni

On Thu\, Jul 04\, 2019 at 12&#8203;:39&#8203;:05PM \-0700\, karl williamson via RT wrote&#8203;:
> On 7/4/19 12&#8203;:01 PM\, Niko Tyni \(via RT\) wrote&#8203;:

> > As discussed in https://bugs.debian.org/931139 and
> > https://bugs.debian.org/924657 glibc has a cache of already loaded gettext
> > translations that gets invalidated in setlocale\(3\)\, bindtextdomain\(3\)
> > and textdomain\(3\) but not uselocale\(3\)\.

> > While this has resulted in a 5\.28 regression / behaviour change for Perl\,
> > it's not clear to me whether glibc is working correctly here or not\. Is
> > this something that Perl should or could work around somehow?
> 
> This looks like an oversight in glibc to me\, hence their bug\.  No matter 
> how the locale gets changed\, the cache should be invalidated\.

Thanks\. I suppose I need to look at reporting this there next\.
 
> In looking at the links you provided\, I didn't see how to read the 
> proposed patch\.

It just adds extra calls to bindtextdomain\(\) and textdomain\(\) to
the application as a workaround\.

Hope these links work for you&#8203;:

&nbsp;  https://salsa.debian.org/installer-team/console-setup/merge_requests/2/diffs

&nbsp;  https://salsa.debian.org/installer-team/console-setup/commit/0f2fc1e85a3a1d316aa28af2fe4f32e027819972

> We should look for workarounds in case glibc refuses to change \(I don't 
> have respect for their willingness to acknowledge flaws in their code\)\, 
> or for working on earlier versions than any fix they do make\.  I wonder 
> if calling foo=textdomain\(NULL\);textdomain\(foo\);
> would do it\.

I had a go with the attached crude patch against 5\.28\.1\, and it does
fix my test case without breaking anything in the main test suite\.

Obviously the textdomain\(\) calls and the inclusion of \<libintl\.h> still
need to be made conditional \(on glibc?\)\, so I guess a new Configure probe
is needed\.

Are there thread safety concerns here? Do we need to worry about one
thread changing the default textdomain right in the middle of another
thread changing locales?
\-\- 
Niko

**Comment by p5pRT** [comment]
### From @ntyni

<details><summary>gettext-caching.diff</summary>

```diff
From 2a410d595ef74bf36c0ea5d06a7afbf59e5d485f Mon Sep 17 00:00:00 2001
From: Niko Tyni <ntyni@debian.org>
Date: Thu, 4 Jul 2019 23:57:59 +0300
Subject: Invalidate glibc translation cache when switching locales

Bug: https://rt.perl.org/Public/Bug/Display.html?id=134264
Bug-Debian: https://bugs.debian.org/931139
Patch-Name: fixes/gettext-caching.diff
---
 locale.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/locale.c b/locale.c
index aedd7da0c..183fd9bdf 100644
--- a/locale.c
+++ b/locale.c
@@ -54,6 +54,8 @@
 #  include <wchar.h>
 #endif
 
+#include <libintl.h>
+
 /* If the environment says to, we can output debugging information during
  * initialization.  This is done before option parsing, and before any thread
  * creation, so can be a file-level static */
@@ -1128,6 +1130,9 @@ S_emulate_setlocale(const int category,
 
 #  endif
 
+    /* Invalidate glibc cache of loaded translations, see [perl #134264] */
+    textdomain(textdomain(NULL));
+
     return locale;
 }
 

```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 7/5/19 1&#8203;:00 AM\, Niko Tyni wrote&#8203;:
> On Thu\, Jul 04\, 2019 at 12&#8203;:39&#8203;:05PM \-0700\, karl williamson via RT wrote&#8203;:
>> On 7/4/19 12&#8203;:01 PM\, Niko Tyni \(via RT\) wrote&#8203;:
> 
>>> As discussed in https://bugs.debian.org/931139 and
>>> https://bugs.debian.org/924657 glibc has a cache of already loaded gettext
>>> translations that gets invalidated in setlocale\(3\)\, bindtextdomain\(3\)
>>> and textdomain\(3\) but not uselocale\(3\)\.
> 
>>> While this has resulted in a 5\.28 regression / behaviour change for Perl\,
>>> it's not clear to me whether glibc is working correctly here or not\. Is
>>> this something that Perl should or could work around somehow?
>>
>> This looks like an oversight in glibc to me\, hence their bug\.  No matter
>> how the locale gets changed\, the cache should be invalidated\.
> 
> Thanks\. I suppose I need to look at reporting this there next\.
>   
>> In looking at the links you provided\, I didn't see how to read the
>> proposed patch\.
> 
> It just adds extra calls to bindtextdomain\(\) and textdomain\(\) to
> the application as a workaround\.
> 
> Hope these links work for you&#8203;:
> 
>    https://salsa.debian.org/installer-team/console-setup/merge_requests/2/diffs
> 
>    https://salsa.debian.org/installer-team/console-setup/commit/0f2fc1e85a3a1d316aa28af2fe4f32e027819972

Thanks
> 
>> We should look for workarounds in case glibc refuses to change \(I don't
>> have respect for their willingness to acknowledge flaws in their code\)\,
>> or for working on earlier versions than any fix they do make\.  I wonder
>> if calling foo=textdomain\(NULL\);textdomain\(foo\);
>> would do it\.
> 
> I had a go with the attached crude patch against 5\.28\.1\, and it does
> fix my test case without breaking anything in the main test suite\.
> 
> Obviously the textdomain\(\) calls and the inclusion of \<libintl\.h> still
> need to be made conditional \(on glibc?\)\, so I guess a new Configure probe
> is needed\.
> 
> Are there thread safety concerns here? Do we need to worry about one
> thread changing the default textdomain right in the middle of another
> thread changing locales?
> 

I don't know\.  The man pages mention nothing about threads\.  That could 
mean an issue\, or it could mean its and non\-issue\.  My guess is that the 
textdomain is global to all threads\.  If the actual change is atomic\, 
then it would just mean that any time a thread did a uselocale\, that it 
would get invalidated\, and so this could happen lots\.  If it isn't 
atomic\, I would expect segfaults\.

Maybe someone in glibc land would know\, or someone could look at the source\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On 7/5/19 10&#8203;:18 PM\, Karl Williamson wrote&#8203;:
> On 7/5/19 1&#8203;:00 AM\, Niko Tyni wrote&#8203;:
>> On Thu\, Jul 04\, 2019 at 12&#8203;:39&#8203;:05PM \-0700\, karl williamson via RT wrote&#8203;:
>>> On 7/4/19 12&#8203;:01 PM\, Niko Tyni \(via RT\) wrote&#8203;:
>>
>>>> As discussed in https://bugs.debian.org/931139 and
>>>> https://bugs.debian.org/924657 glibc has a cache of already loaded 
>>>> gettext
>>>> translations that gets invalidated in setlocale\(3\)\, bindtextdomain\(3\)
>>>> and textdomain\(3\) but not uselocale\(3\)\.
>>
>>>> While this has resulted in a 5\.28 regression / behaviour change for 
>>>> Perl\,
>>>> it's not clear to me whether glibc is working correctly here or not\. Is
>>>> this something that Perl should or could work around somehow?
>>>
>>> This looks like an oversight in glibc to me\, hence their bug\.  No matter
>>> how the locale gets changed\, the cache should be invalidated\.
>>
>> Thanks\. I suppose I need to look at reporting this there next\.
>>> In looking at the links you provided\, I didn't see how to read the
>>> proposed patch\.
>>
>> It just adds extra calls to bindtextdomain\(\) and textdomain\(\) to
>> the application as a workaround\.
>>
>> Hope these links work for you&#8203;:
>>
>>    
>> https://salsa.debian.org/installer-team/console-setup/merge_requests/2/diffs 
>>
>>
>>    
>> https://salsa.debian.org/installer-team/console-setup/commit/0f2fc1e85a3a1d316aa28af2fe4f32e027819972 
>>
> 
> Thanks
>>
>>> We should look for workarounds in case glibc refuses to change \(I don't
>>> have respect for their willingness to acknowledge flaws in their code\)\,
>>> or for working on earlier versions than any fix they do make\.  I wonder
>>> if calling foo=textdomain\(NULL\);textdomain\(foo\);
>>> would do it\.
>>
>> I had a go with the attached crude patch against 5\.28\.1\, and it does
>> fix my test case without breaking anything in the main test suite\.
>>
>> Obviously the textdomain\(\) calls and the inclusion of \<libintl\.h> still
>> need to be made conditional \(on glibc?\)\, so I guess a new Configure probe
>> is needed\.
>>
>> Are there thread safety concerns here? Do we need to worry about one
>> thread changing the default textdomain right in the middle of another
>> thread changing locales?
>>
> 
> I don't know\.  The man pages mention nothing about threads\.  That could 
> mean an issue\, or it could mean its and non\-issue\.  My guess is that the 
> textdomain is global to all threads\.  If the actual change is atomic\, 
> then it would just mean that any time a thread did a uselocale\, that it 
> would get invalidated\, and so this could happen lots\.  If it isn't 
> atomic\, I would expect segfaults\.
> 
> Maybe someone in glibc land would know\, or someone could look at the 
> source\.
> 

And we do have a locale semaphore to solve races with these\.

**Comment by p5pRT** [comment]
### From @khwilliamson

On Sat\, 06 Jul 2019 08&#8203;:44&#8203;:13 \-0700\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 7/5/19 10&#8203;:18 PM\, Karl Williamson wrote&#8203;:
> > On 7/5/19 1&#8203;:00 AM\, Niko Tyni wrote&#8203;:
> >> On Thu\, Jul 04\, 2019 at 12&#8203;:39&#8203;:05PM \-0700\, karl williamson via RT
> >> wrote&#8203;:
> >>> On 7/4/19 12&#8203;:01 PM\, Niko Tyni \(via RT\) wrote&#8203;:
> >>
> >>>> As discussed in https://bugs.debian.org/931139 and
> >>>>  https://bugs.debian.org/924657 glibc has a cache of already
> >>>> loaded
> >>>> gettext
> >>>> translations that gets invalidated in setlocale\(3\)\,
> >>>> bindtextdomain\(3\)
> >>>> and textdomain\(3\) but not uselocale\(3\)\.
> >>
> >>>> While this has resulted in a 5\.28 regression / behaviour change
> >>>> for
> >>>> Perl\,
> >>>> it's not clear to me whether glibc is working correctly here or
> >>>> not\. Is
> >>>> this something that Perl should or could work around somehow?
> >>>
> >>> This looks like an oversight in glibc to me\, hence their bug\.  No
> >>> matter
> >>> how the locale gets changed\, the cache should be invalidated\.
> >>
> >> Thanks\. I suppose I need to look at reporting this there next\.
> >>> In looking at the links you provided\, I didn't see how to read the
> >>> proposed patch\.
> >>
> >> It just adds extra calls to bindtextdomain\(\) and textdomain\(\) to
> >> the application as a workaround\.
> >>
> >> Hope these links work for you&#8203;:
> >>
> >>
> >> https://salsa.debian.org/installer-team/console-
> >> setup/merge\_requests/2/diffs
> >>
> >>
> >>
> >> https://salsa.debian.org/installer-team/console-
> >> setup/commit/0f2fc1e85a3a1d316aa28af2fe4f32e027819972
> >>
> >
> > Thanks
> >>
> >>> We should look for workarounds in case glibc refuses to change \(I
> >>> don't
> >>> have respect for their willingness to acknowledge flaws in their
> >>> code\)\,
> >>> or for working on earlier versions than any fix they do make\.  I
> >>> wonder
> >>> if calling foo=textdomain\(NULL\);textdomain\(foo\);
> >>> would do it\.
> >>
> >> I had a go with the attached crude patch against 5\.28\.1\, and it does
> >> fix my test case without breaking anything in the main test suite\.
> >>
> >> Obviously the textdomain\(\) calls and the inclusion of \<libintl\.h>
> >> still
> >> need to be made conditional \(on glibc?\)\, so I guess a new Configure
> >> probe
> >> is needed\.
> >>
> >> Are there thread safety concerns here? Do we need to worry about one
> >> thread changing the default textdomain right in the middle of
> >> another
> >> thread changing locales?
> >>
> >
> > I don't know\.  The man pages mention nothing about threads\.  That
> > could
> > mean an issue\, or it could mean its and non\-issue\.  My guess is that
> > the
> > textdomain is global to all threads\.  If the actual change is atomic\,
> > then it would just mean that any time a thread did a uselocale\, that
> > it
> > would get invalidated\, and so this could happen lots\.  If it isn't
> > atomic\, I would expect segfaults\.
> >
> > Maybe someone in glibc land would know\, or someone could look at the
> > source\.
> >
> 
> And we do have a locale semaphore to solve races with these\.

Ticket against glibc filed&#8203;: https://sourceware.org/bugzilla/show_bug.cgi?id=24936
\-\- 
Karl Williamson

**Comment by khwilliamson** [comment]
@ntyni  Do you want perl to change to invalidate the cache under gcc

**Comment by ntyni** [comment]
On Mon, Nov 25, 2019 at 04:47:45PM -0800, Karl Williamson wrote:
> @ntyni  Do you want perl to change to invalidate the cache under gcc

I guess you mean under glibc.

The perl behaviour regressed in 5.28 due to this, and glibc
doesn't look like it's getting fixed any time soon.

So yes, having a workaround in perl would be nice.

Thanks for looking at this,
-- 
Niko

**Comment by ntyni** [comment]
Confirming that your workaround works for me as well, and passes the test suite here.

**Comment by khwilliamson** [comment]
Note that you can Configure with `-Accflags=PERL_NO_2008_LOCALE` to avoid this bug.  I don't remember if this symbol is available in 5.28, or not until 5.30.

I have created a branch for your perusal which I believe fixes this bug, smoke-me/khw-locale.  There is already a symbol `__GLIBC__ ` used in a bunch of places in the core.  I believe that it has the obvious meaning, so no new Configure probe is necessary.  I tested this branch on your test script you furnished, and I get three different texts as in the pre 5.28 case.  Two of the languages I can read, and they have the correct text in them.  You can pretty much guess which one I can't read. 

My patch basically just fancies up yours, the one you called crude, and makes sure that it doesn't invalidate the cache if some other category instead of `LC_MESSAGES` is changed.

I think we'll know that there is a thread race if there start to be segfault reports.  Those can be solved, I believe by use of the existing semaphore.  We won't know if there is such a race if no reports come in, unless we get some assurance from the glibc folks.  (Don't laugh so hard.)
